### PR TITLE
Number extractor service

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | A service to extract Number() object from any given object or array. It uses symfony property accessor inside to reach the properties by passing path as string. We need to use Number in many calculations, this should save some time on the convertion of ObjectModel floats to Number().
| Type?         |  refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | There is nothing to test in UI. Added couple unit tests that should pass with CI

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/19648)

### CODE REVIEW

File: src/Core/Util/Number/NumberExtractor.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+namespace PrestaShop\PrestaShop\Core\Util\Number;
+
+use PrestaShop\Decimal\Number;
+use Symfony\Component\PropertyAccess\Exception\AccessException;
+use Symfony\Component\PropertyAccess\Exception\InvalidArgumentException;
+use Symfony\Component\PropertyAccess\Exception\UnexpectedTypeException;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * Extracts numeric value as @var Number from given resource
```
Comment (matks) [inline src/Core/Util/Number/NumberExtractor.php:39]: A little example would be great inside this phpDoc 😄

File: src/Core/Util/Number/NumberExtractor.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    private $propertyAccessor;
+
+    public function __construct(PropertyAccessorInterface $propertyAccessor)
+    {
+        $this->propertyAccessor = $propertyAccessor;
+    }
+
+    /**
+     * If provided resource is array, access its values using brackets e.g. '[one_property][another_level_property]'
```
Comment (jolelievre) [inline src/Core/Util/Number/NumberExtractor.php:54]: I thought the property accessor can use both syntax indiferently? Like `one_property.another_level_property` can work on an array no?

Comment (zuk3975) [reply src/Core/Util/Number/NumberExtractor.php:54]: I thought so too, but tests shown that it must be accessed with brackets when using array :man_shrugging:

**Review by jolelievre — APPROVED** [review]
PHPDoc might be improved, but can be done later ;)

File: src/PrestaShopBundle/Resources/config/services/core/util/number.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+services:
+  _defaults:
+    public: true
+
+  prestashop.core.util.number.number_extractor:
+    class: 'PrestaShop\PrestaShop\Core\Util\Number\NumberExtractor'
+    arguments:
+      '@property_accessor'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/core/util/number.yml]: Oh that's why travis fails!
```suggestion
      - '@property_accessor'
```

**Review by jolelievre — CHANGES_REQUESTED** [review]

**Review by jolelievre — APPROVED** [review]

**Comment by zuk3975** [comment]
rr travis

**Review by PierreRambaud — APPROVED** [review]

**Comment by matks** [comment]
Thank you @zuk3975

**Pull request merged**